{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyP2ZjmKyBXQ1E37jqpq8stK"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**BIOPYTHON**\n","\n","- Biopython Project is an international association of developers of freely available Python (https://www.python.org) tools for computational molecular biology.\n","- The Biopython web site (http://www.biopython.org) provides an online resource for modules, scriptsand web links for developers of Python-based software for bioinformatics use and research.\n","\n","----"],"metadata":{"id":"GHtdXuTMWEUp"}},{"cell_type":"markdown","source":["Install Biopython"],"metadata":{"id":"hjzZZoZKW4DE"}},{"cell_type":"code","source":["pip install biopython"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Ci5kQbSfWllu","executionInfo":{"status":"ok","timestamp":1732694873912,"user_tz":-330,"elapsed":6337,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"304339ef-03e1-42f7-b925-f92e4705bcf0"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.26.4)\n","Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m24.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: biopython\n","Successfully installed biopython-1.84\n"]}]},{"cell_type":"markdown","source":["verify installation"],"metadata":{"id":"1Y15--QsW8gt"}},{"cell_type":"code","source":["import Bio\n","print(\"Biopython version:\", Bio.__version__)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Fgrkq2qxXCTR","executionInfo":{"status":"ok","timestamp":1732694877659,"user_tz":-330,"elapsed":507,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"70f79c76-d059-452b-c596-d3001bb35dcf"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["Biopython version: 1.84\n"]}]},{"cell_type":"markdown","source":["Biopyhton provides a wide range of modules designed to handle various biological data types and perform common bioinformatics tasks."],"metadata":{"id":"E8og8LByXEfk"}},{"cell_type":"markdown","source":["\n","### **SEQUENCE**\n"," - If sequences were stored as strings, there would be no way to know what kind of sequence it is. This is why Biopython introduces Seq objects"],"metadata":{"id":"Z6Mg0LhrZLJ6"}},{"cell_type":"markdown","source":["Import seq"],"metadata":{"id":"JVk-nisFkSP4"}},{"cell_type":"code","source":["from Bio.Seq import Seq"],"metadata":{"id":"zP1YsJCCkQAN"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["my_seq = Seq(\"AGTACACTGGT\")\n","my_seq                                                                          # prints the loaded sequence"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"FokHt7wIaVjZ","executionInfo":{"status":"ok","timestamp":1731904777176,"user_tz":-330,"elapsed":494,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"2e42f012-c64e-4de3-f6e9-53f02181ee2d"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["Seq('AGTACACTGGT')"]},"metadata":{},"execution_count":25}]},{"cell_type":"markdown","source":[".count () : count occurrences of a specific nucleotide or subsequence"],"metadata":{"id":"pvUDVycnfcpU"}},{"cell_type":"code","source":["count_A = my_seq.count(\"A\")                                                     # Count occurrences of 'A'\n","count_C = my_seq.count(\"C\")                                                     # Count occurrences of 'C'\n","count_T = my_seq.count(\"T\")                                                     # Count occurrences of 'T'\n","count_G = my_seq.count(\"G\")                                                     # Count occurrences of 'G'\n","\n","                                                                                # Print the results\n","print(f\"Count of 'A': {count_A}\")\n","print(f\"Count of 'C': {count_C}\")\n","print(f\"Count of 'T': {count_T}\")\n","print(f\"Count of 'G': {count_G}\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"zUrv5jNUfb3T","executionInfo":{"status":"ok","timestamp":1731904780950,"user_tz":-330,"elapsed":857,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"4fef3e68-247d-43d1-d227-0fc4cb31bca6"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Count of 'A': 3\n","Count of 'C': 2\n","Count of 'T': 3\n","Count of 'G': 3\n"]}]},{"cell_type":"markdown","source":["Slicing a sequence"],"metadata":{"id":"yEI7STx2gTVj"}},{"cell_type":"code","source":["my_seq = Seq(\"GATCGATGGGCCTATATAGGATCGAAAATCGC\")\n","\n","slice_1 = my_seq[0:5]                                                           # Get the first five characters\n","slice_2 = my_seq[5:10]                                                          # Get characters from index 5 to 9\n","slice_3 = my_seq[::2]                                                           # Get every second character (stride of 2)\n","slice_4 = my_seq[-3:]                                                           # Get the last three characters\n","\n","                                                                                # Print the results\n","print(\"Original Sequence: \", my_seq)\n","print(\"Slice 1 (0:5): \", slice_1)                                               # Output: AGTAC\n","print(\"Slice 2 (5:10): \", slice_2)                                              # Output: ACTGG\n","print(\"Slice 3 (::2): \", slice_3)                                               # Output: ATCTG\n","print(\"Slice 4 (-3:): \", slice_4)                                               # Output: GGT"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"collapsed":true,"id":"TEVQbf3ogog2","executionInfo":{"status":"ok","timestamp":1731904805551,"user_tz":-330,"elapsed":504,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"bbd34515-b2ed-4c45-d17f-77d865c75049"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Original Sequence:  GATCGATGGGCCTATATAGGATCGAAAATCGC\n","Slice 1 (0:5):  GATCG\n","Slice 2 (5:10):  ATGGG\n","Slice 3 (::2):  GTGTGCTTTGACAATG\n","Slice 4 (-3:):  CGC\n"]}]},{"cell_type":"markdown","source":["concatenate or add sequences"],"metadata":{"id":"iYwWjThfiHc7"}},{"cell_type":"code","source":["seq1 = Seq(\"ACGT\")\n","seq2 = Seq(\"AACCGG\")\n","\n","combined_seq = seq1 + seq2                                                      # Concatenates the sequences using the + operator\n","\n","print(\"Combined Sequence:\", combined_seq)                                       # prints the output: Seq('ACGTAACCGG')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"8UDnGDJJiP-P","executionInfo":{"status":"ok","timestamp":1731904808899,"user_tz":-330,"elapsed":513,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"74646e4e-6324-47c5-bb39-87b210ccf119"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Combined Sequence: ACGTAACCGG\n"]}]},{"cell_type":"code","source":["list_of_seqs = [Seq(\"ACGT\"), Seq(\"AACC\"), Seq(\"GGTT\"), Seq(\"ATGCT\")]            # Creates a list of Seq objects\n","\n","concatenated = Seq(\"\")                                                          # Initialize an empty Seq object for concatenation\n","\n","for s in list_of_seqs:                                                          # Concatenate each sequence in the list using a for loop\n","    concatenated += s\n","\n","print(\"The Sequence is :\", concatenated)                                        # prints the concatenated output: Seq('ACGTAACCGGTT')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"-zORRVz5kziQ","executionInfo":{"status":"ok","timestamp":1731905048102,"user_tz":-330,"elapsed":608,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"4a3f3329-2d9a-49a2-8e19-dffe00bce9de"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["The Sequence is : ACGTAACCGGTTATGCT\n"]}]},{"cell_type":"markdown","source":["Nucleotide sequences and (reverse) complements"],"metadata":{"id":"SFJOwTxdyapA"}},{"cell_type":"code","source":["my_seq = Seq(\"GATCGATGGGCCTATATAGGATCGAAAATCGC\")\n","print('The original seq is : \\n', my_seq)                                       # prints the original sequence\n","\n","print('The complement of the seq is : \\n', my_seq.complement())                 # prints the complement of the input sequence\n","\n","print('The reverse of the complement seq is : \\n',my_seq.reverse_complement())  # prints the reverse of the complement"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"y_VWksLMybyf","executionInfo":{"status":"ok","timestamp":1731908883691,"user_tz":-330,"elapsed":560,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"1b0fea9b-5155-4fc2-ca5e-2c393964af1c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["The original seq is : \n"," GATCGATGGGCCTATATAGGATCGAAAATCGC\n","The complement of the seq is : \n"," CTAGCTACCCGGATATATCCTAGCTTTTAGCG\n","The reverse of the complement seq is : \n"," GCGATTTTCGATCCTATATAGGCCCATCGATC\n"]}]},{"cell_type":"markdown","source":["Transcription"],"metadata":{"id":"n_lvaZt40RZm"}},{"cell_type":"code","source":["dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n","print ('The DNA Sequence is: \\n', dna)\n","\n","mrna = dna.transcribe()                                                         #switches T → U\n","print ('The messenger RNA of the input DNA is: \\n', mrna)\n","\n","t_mrna = mrna.back_transcribe()                                                 #back-transcription method for going from the mRNA to the coding strand of the DNA.\n","print('The reverse transcribed sequence of mRNA is : \\n', t_mrna)"],"metadata":{"id":"7it262xT0SOm"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Translation Tables\n","\n","- The translation method uses codon table objects derived from\n","the NCBI information at ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt, also shown on https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi in a much more readable layout.\n","- There are 33 transaltional tables available in the above link.\n","- Each table has a name as well as numbers assigned  \n"," - Example: the 1st table is the \"standard code\" mentioned as \"trans_table = 1\"."],"metadata":{"id":"6Ytn03zMGzo5"}},{"cell_type":"code","source":["from Bio.Data import CodonTable\n","mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"]     #import the 2nd table\n","\n","#to import the tables using the number assigned\n","#mito_table = CodonTable.unambiguous_dna_by_id[2]\n","\n","print(mito_table)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"UEKjDGq-G3K4","executionInfo":{"status":"ok","timestamp":1731918882700,"user_tz":-330,"elapsed":627,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"d67597b4-078f-4c45-a5fc-74d59186b58e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Table 2 Vertebrate Mitochondrial, SGC1\n","\n","  |  T      |  C      |  A      |  G      |\n","--+---------+---------+---------+---------+--\n","T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n","T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n","T | TTA L   | TCA S   | TAA Stop| TGA W   | A\n","T | TTG L   | TCG S   | TAG Stop| TGG W   | G\n","--+---------+---------+---------+---------+--\n","C | CTT L   | CCT P   | CAT H   | CGT R   | T\n","C | CTC L   | CCC P   | CAC H   | CGC R   | C\n","C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n","C | CTG L   | CCG P   | CAG Q   | CGG R   | G\n","--+---------+---------+---------+---------+--\n","A | ATT I(s)| ACT T   | AAT N   | AGT S   | T\n","A | ATC I(s)| ACC T   | AAC N   | AGC S   | C\n","A | ATA M(s)| ACA T   | AAA K   | AGA Stop| A\n","A | ATG M(s)| ACG T   | AAG K   | AGG Stop| G\n","--+---------+---------+---------+---------+--\n","G | GTT V   | GCT A   | GAT D   | GGT G   | T\n","G | GTC V   | GCC A   | GAC D   | GGC G   | C\n","G | GTA V   | GCA A   | GAA E   | GGA G   | A\n","G | GTG V(s)| GCG A   | GAG E   | GGG G   | G\n","--+---------+---------+---------+---------+--\n"]}]},{"cell_type":"code","source":["start = mito_table.start_codons                                                 #the start codons from the table are selected\n","print ('The start codons of the mito table are : \\n', start)\n","\n","stop = mito_table.stop_codons                                                   #the stop codons from the table are selected\n","print ('The stop codons are of the mito table : \\n', stop)"],"metadata":{"id":"ERknMGavIUMv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Translation"],"metadata":{"id":"nHh8QLif0Sln"}},{"cell_type":"code","source":["T_dna = dna.translate()\n","print('The translated Seq of the input DNA is : \\n', T_dna)                     # the * in the output indicates the STOP CODON.\n","\n","Tr_dna = dna.translate(to_stop=True)                                            #stops just before the 1st stop codon\n","                                                                                #and the stop symbol is not included at the end of the protein sequence.\n","print('The translated Seq without stop codon of the input DNA is : \\n', Tr_dna)\n","\n","mito_T_dna = dna.translate(table=2, to_stop = True)\n","print('The translated Seq of the input DNA using mito table : \\n', mito_T_dna)  #compares with the table for the stop codons and does not give * if the stop codon occurs somewhere in the middle of the seq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kcC3xilO0Um3","executionInfo":{"status":"ok","timestamp":1731919152522,"user_tz":-330,"elapsed":522,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"593602ad-2e72-4819-9943-3e73b0befb98"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["The translated Seq of the input DNA is : \n"," MAIVMGR*KGAR*\n","The translated Seq without stop codon of the input DNA is : \n"," MAIVMGR\n","The translated Seq of the input DNA using mito table : \n"," MAIVMGRWKGAR\n"]}]},{"cell_type":"markdown","source":["Comparing Seq objects"],"metadata":{"id":"jbTnH4qn0VB_"}},{"cell_type":"code","source":["seq1 = Seq(\"AGTACACTGGT\")                                                       # Create two Seq objects\n","seq2 = Seq(\"AGTACACTGGT\")\n","seq3 = Seq(\"AGTACGTTGGT\")\n","seq4 = Seq(\"AGTACACTG\")\n","\n","                                                                                # Equality comparison using ==\n","print(\"seq1 == seq2: \\n\", seq1 == seq2)                                         # True\n","print(\"seq1 == seq3: \\n\", seq1 == seq3)                                         # False\n","\n","                                                                                # Inequality comparison using !=\n","print(\"seq1 != seq3: \\n\", seq1 != seq3)                                         # True\n","\n","                                                                                # Lexicographical comparison using < or > or <= or >=\n","print(\"seq1 < seq3: \\n\", seq1 < seq3)                                           # True (compares lexicographically)\n","print(\"seq1 > seq4: \\n\", seq1 > seq4)                                           # True (since length matters in lexicographical order)\n","\n","# Length comparison\n","print(\"Length of seq1: \\n\", len(seq1))                                          # Output: 11\n","print(\"Length of seq4: \\n\", len(seq4))                                          # Output: 10"],"metadata":{"id":"QqyentNT0ZGO","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1731919648982,"user_tz":-330,"elapsed":827,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"38119194-b391-444a-fc43-52d90c63206c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["seq1 == seq2: \n"," True\n","seq1 == seq3: \n"," False\n","seq1 != seq3: \n"," True\n","seq1 < seq3: \n"," True\n","seq1 > seq4: \n"," True\n","Length of seq1: \n"," 11\n","Length of seq4: \n"," 9\n"]}]},{"cell_type":"markdown","source":["###**Sequence Input/Output**"],"metadata":{"id":"uBC4m4ONfgoS"}},{"cell_type":"markdown","source":["Reading Sequence Files"],"metadata":{"id":"QEchS3C2fulI"}},{"cell_type":"markdown","source":["FASTA\n"],"metadata":{"id":"eIqMJN2EcqWM"}},{"cell_type":"code","source":["from Bio import SeqIO\n","for seq_record in SeqIO.parse(\"ls_orchid.fasta\", \"fasta\"):                      #loads the sequence in fasta format\n","  print(seq_record.id)\n","  print(repr(seq_record.seq))\n","  print(len(seq_record))"],"metadata":{"id":"TEhA5C5Eaj-6"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["GenBank"],"metadata":{"id":"E42zT24MczSz"}},{"cell_type":"code","source":["from Bio import SeqIO\n","for seq_record in SeqIO.parse(\"ls_orchid.gbk\", \"genbank\"):                      #loads the sequence in genbank format\n","  print(seq_record.id)\n","  print(repr(seq_record.seq))\n","  print(len(seq_record))"],"metadata":{"id":"X-22ywrfc7B7","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732695050072,"user_tz":-330,"elapsed":812,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"f3726358-6f8c-4e56-8f51-66d9aedfc16d"},"execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["Z78533.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC')\n","740\n","Z78532.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAACAG...GGC')\n","753\n","Z78531.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TAA')\n","748\n","Z78530.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAAACAACAT...CAT')\n","744\n","Z78529.1\n","Seq('ACGGCGAGCTGCCGAAGGACATTGTTGAGACAGCAGAATATACGATTGAGTGAA...AAA')\n","733\n","Z78527.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...CCC')\n","718\n","Z78526.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...TGT')\n","730\n","Z78525.1\n","Seq('TGTTGAGATAGCAGAATATACATCGAGTGAATCCGGAGGACCTGTGGTTATTCG...GCA')\n","704\n","Z78524.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATAGTAG...AGC')\n","740\n","Z78523.1\n","Seq('CGTAACCAGGTTTCCGTAGGTGAACCTGCGGCAGGATCATTGTTGAGACAGCAG...AAG')\n","709\n","Z78522.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...GAG')\n","700\n","Z78521.1\n","Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAGAATATATGATCGAGT...ACC')\n","726\n","Z78520.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TTT')\n","753\n","Z78519.1\n","Seq('ATATGATCGAGTGAATCTGGTGGACTTGTGGTTACTCAGCTCGCCATAGGCTTT...TTA')\n","699\n","Z78518.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATAGTAG...TCC')\n","658\n","Z78517.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...AGC')\n","752\n","Z78516.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAT...TAA')\n","726\n","Z78515.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGCTGAGACCGTAG...AGC')\n","765\n","Z78514.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...CTA')\n","755\n","Z78513.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GAG')\n","742\n","Z78512.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...AGC')\n","762\n","Z78511.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCCCCA...GGA')\n","745\n","Z78510.1\n","Seq('CTAACCAGGGTTCCGAGGTGACCTTCGGGAGGATTCCTTTTTAAGCCCCCGAAA...TTA')\n","750\n","Z78509.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GGA')\n","731\n","Z78508.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TGA')\n","741\n","Z78507.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCCCCA...TGA')\n","740\n","Z78506.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...TGA')\n","727\n","Z78505.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TTT')\n","711\n","Z78504.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCGCAA...TAA')\n","743\n","Z78503.1\n","Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCCTTGTTGAGACCGCCA...TAA')\n","727\n","Z78502.1\n","Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGACCGCCA...CGC')\n","757\n","Z78501.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...AGA')\n","770\n","Z78500.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGCTCATTGTTGAGACCGCAA...AAG')\n","767\n","Z78499.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAGGGATCATTGTTGAGATCGCAT...ACC')\n","759\n","Z78498.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAAGGTCATTGTTGAGATCACAT...AGC')\n","750\n","Z78497.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n","788\n","Z78496.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC')\n","774\n","Z78495.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GTG')\n","789\n","Z78494.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGGTCGCAT...AAG')\n","688\n","Z78493.1\n","Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...CCC')\n","719\n","Z78492.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...ATA')\n","743\n","Z78491.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC')\n","737\n","Z78490.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n","728\n","Z78489.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGC')\n","740\n","Z78488.1\n","Seq('CTGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGCAATAATTGATCGA...GCT')\n","696\n","Z78487.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TAA')\n","732\n","Z78486.1\n","Seq('CGTCACGAGGTTTCCGTAGGTGAATCTGCGGGAGGATCATTGTTGAGATCACAT...TGA')\n","731\n","Z78485.1\n","Seq('CTGAACCTGGTGTCCGAAGGTGAATCTGCGGATGGATCATTGTTGAGATATCAT...GTA')\n","735\n","Z78484.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGGGGAAGGATCATTGTTGAGATCACAT...TTT')\n","720\n","Z78483.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n","740\n","Z78482.1\n","Seq('TCTACTGCAGTGACCGAGATTTGCCATCGAGCCTCCTGGGAGCTTTCTTGCTGG...GCA')\n","629\n","Z78481.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n","572\n","Z78480.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n","587\n","Z78479.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGT')\n","700\n","Z78478.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCAGTGTTGAGATCACAT...GGC')\n","636\n","Z78477.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n","716\n","Z78476.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC')\n","592\n","Z78475.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT')\n","716\n","Z78474.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGT...CTT')\n","733\n","Z78473.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n","626\n","Z78472.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n","737\n","Z78471.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n","740\n","Z78470.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT')\n","574\n","Z78469.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT')\n","594\n","Z78468.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...GTT')\n","610\n","Z78467.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n","730\n","Z78466.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC')\n","641\n","Z78465.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n","702\n","Z78464.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAGCGGAAGGGTCATTGTTGAGATCACATAATA...AGC')\n","733\n","Z78463.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGTTCATTGTTGAGATCACAT...AGC')\n","738\n","Z78462.1\n","Seq('CGTCACGAGGTCTCCGGATGTGACCCTGCGGAAGGATCATTGTTGAGATCACAT...CAT')\n","736\n","Z78461.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TAA')\n","732\n","Z78460.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TTA')\n","745\n","Z78459.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTT')\n","744\n","Z78458.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTG')\n","738\n","Z78457.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GAG')\n","739\n","Z78456.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n","740\n","Z78455.1\n","Seq('CGTAACCAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAGATCACAT...GCA')\n","745\n","Z78454.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AAC')\n","695\n","Z78453.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n","745\n","Z78452.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n","743\n","Z78451.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGTACCTCCGGAAGGATCATTGTTGAGATCACAT...AGC')\n","730\n","Z78450.1\n","Seq('GGAAGGATCATTGCTGATATCACATAATAATTGATCGAGTTAAGCTGGAGGATC...GAG')\n","706\n","Z78449.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n","744\n","Z78448.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n","742\n","Z78447.1\n","Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATCACAT...AGC')\n","694\n","Z78446.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...CCC')\n","712\n","Z78445.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGT')\n","715\n","Z78444.1\n","Seq('CGTAACAAGGTTTCCGTAGGGTGAACTGCGGAAGGATCATTGTTGAGATCACAT...ATT')\n","688\n","Z78443.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n","784\n","Z78442.1\n","Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACATAATAATTGATCGAGT...AGT')\n","721\n","Z78441.1\n","Seq('GGAAGGTCATTGCCGATATCACATAATAATTGATCGAGTTAATCTGGAGGATCT...GAG')\n","703\n","Z78440.1\n","Seq('CGTAACAAGGTTTCCGTAGGTGGACCTCCGGGAGGATCATTGTTGAGATCACAT...GCA')\n","744\n","Z78439.1\n","Seq('CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC')\n","592\n"]}]},{"cell_type":"markdown","source":["Converting Seq file formats"],"metadata":{"id":"KnRh3_Pff3V-"}},{"cell_type":"code","source":["from Bio import SeqIO\n","with open(\"ls_orchid.gbk\") as input_handle, open(\"my_example.fasta\", \"w\") as output_handle:\n","  sequences = SeqIO.parse(input_handle, \"genbank\")\n","  count = SeqIO.write(sequences, output_handle, \"fasta\")\n","print(\"Converted %i records\" % count)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"f9Pk47ThrnTT","executionInfo":{"status":"ok","timestamp":1732695293237,"user_tz":-330,"elapsed":487,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"88f75c2e-0520-4c02-a5b7-8436bcf8f3b5"},"execution_count":11,"outputs":[{"output_type":"stream","name":"stdout","text":["Converted 94 records\n"]}]},{"cell_type":"markdown","source":["## **MULTIPLE SEQUENCE ALIGNMENT**"],"metadata":{"id":"DWiWiTbQsRur"}},{"cell_type":"markdown","source":["Single Alignments\n","\n","- consider the following annotation rich protein alignment in the PFAM or Stockholm file format:"],"metadata":{"id":"Yf4CGpuBsaTD"}},{"cell_type":"code","source":["from Bio import AlignIO\n","alignment = AlignIO.read(\"PF05371_seed.sth\", \"stockholm\")\n","print(alignment)                                                                # summary of the alignment"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QUboQdxktyeq","executionInfo":{"status":"ok","timestamp":1732695820814,"user_tz":-330,"elapsed":658,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"f37951f4-fa55-4f11-afbc-201c637bb544"},"execution_count":13,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 52 columns\n","AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRL...SKA COATB_BPIKE/30-81\n","AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKL...SRA Q9T0Q8_BPIKE/1-52\n","DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRL...SKA COATB_BPI22/32-83\n","AEGDDP---AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA COATB_BPM13/24-72\n","AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA COATB_BPZJ2/1-49\n","AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA Q9T0Q9_BPFD/1-49\n","FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL...SRA COATB_BPIF1/22-73\n"]}]},{"cell_type":"markdown","source":["Multiple Alignments"],"metadata":{"id":"7Sk5cozTufWz"}},{"cell_type":"code","source":["\n","from Bio import AlignIO\n","alignments = AlignIO.parse(\"resampled.phy\", \"phylip\")\n","for alignment in alignments:\n","    print(alignment)\n","    print(\"\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"4OH075jbuf0y","executionInfo":{"status":"ok","timestamp":1732695994839,"user_tz":-330,"elapsed":497,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"175344af-5f47-4399-ac71-54877251630b"},"execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 5 rows and 6 columns\n","AAACCA Alpha\n","AAACCC Beta\n","ACCCCA Gamma\n","CCCAAC Delta\n","CCCAAA Epsilon\n","\n","Alignment with 5 rows and 6 columns\n","AAACAA Alpha\n","AAACCC Beta\n","ACCCAA Gamma\n","CCCACC Delta\n","CCCAAA Epsilon\n","\n","Alignment with 5 rows and 6 columns\n","AAAAAC Alpha\n","AAACCC Beta\n","AACAAC Gamma\n","CCCCCA Delta\n","CCCAAC Epsilon\n","\n","Alignment with 5 rows and 6 columns\n","AAAACC Alpha\n","ACCCCC Beta\n","AAAACC Gamma\n","CCCCAA Delta\n","CAAACC Epsilon\n","\n"]}]},{"cell_type":"markdown","source":["Ambiguous Alignment:\n","\n","Many alignment file formats can explicitly store more than one alignment, and the division between each alignment is clear. However, when a general sequence file format has been used there is no such block structure.\n","\n","Clearly trying to store more than one alignment in a FASTA file is not ideal. Bio.AlignIO can cope with the most common situation where all the alignments have the same number of records.\n","\n","We can use Bio.AlignIO.parse() with the optional seq_count argument which specifies how many sequences are expected in each alignment (in these examples, 3, 2 and 2 respectively)."],"metadata":{"id":"abVocPRAvYPL"}},{"cell_type":"code","source":["for alignment in AlignIO.parse(\"handle.phy\", \"fasta\", seq_count=2):\n","    print(\"Alignment length %i\" % alignment.get_alignment_length())\n","    for record in alignment:\n","        print(\"%s - %s\" % (record.seq, record.id))\n","    print(\"\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yJ7q0HqDvvyz","executionInfo":{"status":"ok","timestamp":1732696305221,"user_tz":-330,"elapsed":493,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"7f68077b-139d-4ea2-c6ed-b07f8fc2eeca"},"execution_count":15,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment length 19\n","ACTACGACTAGCTCAG--G - Alpha\n","ACTACCGCTAGCTCAGAAG - XXX\n","\n","Alignment length 17\n","ACTACGACTAGCTCAGG - Alpha\n","ACTACGGCAAGCACAGG - YYY\n","\n","Alignment length 21\n","--ACTACGAC--TAGCTCAGG - Alpha\n","GGACTACGACAATAGCTCAGG - ZZZ\n","\n"]}]},{"cell_type":"markdown","source":["Writing Alignments:\n","\n","Bio.AlignIO.write() is a function taking three arguments: some MultipleSeqAlignment objects (or for backwards compatibility the obsolete Alignment objects), a handle or filename to write to, and a sequence format."],"metadata":{"id":"RoRHRvhIv3nq"}},{"cell_type":"code","source":["from Bio.Seq import Seq\n","from Bio.SeqRecord import SeqRecord\n","from Bio.Align import MultipleSeqAlignment\n","                                                                                #create some SeqRecord objects to construct the alignment\n","align1 = MultipleSeqAlignment([\n","             SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\"),\n","             SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\"),\n","             SeqRecord(Seq(\"ACTGCTAGDTAG\"), id=\"Gamma\"),\n","         ])\n","\n","align2 = MultipleSeqAlignment([\n","             SeqRecord(Seq(\"GTCAGC-AG\"), id=\"Delta\"),\n","             SeqRecord(Seq(\"GACAGCTAG\"), id=\"Epsilon\"),\n","             SeqRecord(Seq(\"GTCAGCTAG\"), id=\"Zeta\"),\n","         ])\n","\n","align3 = MultipleSeqAlignment([\n","             SeqRecord(Seq(\"ACTAGTACAGCTG\"), id=\"Eta\"),\n","             SeqRecord(Seq(\"ACTAGTACAGCT-\"), id=\"Theta\"),\n","             SeqRecord(Seq(\"-CTACTACAGGTG\"), id=\"Iota\"),\n","         ])\n","\n","my_alignments = [align1, align2, align3]\n","\n","from Bio import AlignIO\n","AlignIO.write(my_alignments, \"my_example.phy\", \"phylip\")                        #write to a PHYLIP format file\n","\n","!head my_example.phy                                                            #prints the 1st few rows in the created .phy file"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"dwu48nsjwMPS","executionInfo":{"status":"ok","timestamp":1732696542380,"user_tz":-330,"elapsed":485,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"0299ca61-432f-4fc5-feed-11f2f8a32b9e"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":[" 3 12\n","Alpha      ACTGCTAGCT AG\n","Beta       ACT-CTAGCT AG\n","Gamma      ACTGCTAGDT AG\n"," 3 9\n","Delta      GTCAGC-AG\n","Epsilon    GACAGCTAG\n","Zeta       GTCAGCTAG\n"," 3 13\n","Eta        ACTAGTACAG CTG\n"]}]},{"cell_type":"markdown","source":["Converting between sequence alignment file formats:\n","\n","- Converting between sequence alignment file formats with Bio.AlignIO works in the same way as converting between sequence file formats with Bio.SeqIO"],"metadata":{"id":"KN3rUGv6xGnk"}},{"cell_type":"code","source":["from Bio import AlignIO\n","count = AlignIO.convert(\"PF05371_seed.sth\", \"stockholm\", \"PF05371_seed.aln\", \"clustal\")\n","print(\"Converted %i alignments\" % count)\n","\n","!cat PF05371_seed.aln"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YXWPlLxgxG98","executionInfo":{"status":"ok","timestamp":1732696796146,"user_tz":-330,"elapsed":757,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"fb53b663-4d75-486a-8c0d-d864079d7382"},"execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":["Converted 1 alignments\n","CLUSTAL X (1.81) multiple sequence alignment\n","\n","\n","COATB_BPIKE/30-81                   AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSS\n","Q9T0Q8_BPIKE/1-52                   AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVS\n","COATB_BPI22/32-83                   DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSS\n","COATB_BPM13/24-72                   AEGDDP---AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTS\n","COATB_BPZJ2/1-49                    AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFAS\n","Q9T0Q9_BPFD/1-49                    AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTS\n","COATB_BPIF1/22-73                   FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVS\n","\n","COATB_BPIKE/30-81                   KA\n","Q9T0Q8_BPIKE/1-52                   RA\n","COATB_BPI22/32-83                   KA\n","COATB_BPM13/24-72                   KA\n","COATB_BPZJ2/1-49                    KA\n","Q9T0Q9_BPFD/1-49                    KA\n","COATB_BPIF1/22-73                   RA\n","\n","\n"]}]},{"cell_type":"markdown","source":["#**Manipulating Alignments**"],"metadata":{"id":"4JRwv0KByLHs"}},{"cell_type":"markdown","source":["Slicing alignments"],"metadata":{"id":"fYJmoy37yUfE"}},{"cell_type":"code","source":["from Bio import AlignIO\n","alignment = AlignIO.read(\"PF05371_seed.sth\", \"stockholm\")\n","print(\"Number of rows: %i\" % len(alignment))\n",""],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Df3escu_yVN7","executionInfo":{"status":"ok","timestamp":1732696997025,"user_tz":-330,"elapsed":491,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"2dd2b55c-7123-487e-aa02-42b26274fd9a"},"execution_count":21,"outputs":[{"output_type":"stream","name":"stdout","text":["Number of rows: 7\n"]}]},{"cell_type":"code","source":["print(alignment)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"KOIodrp1y6IT","executionInfo":{"status":"ok","timestamp":1732698504704,"user_tz":-330,"elapsed":615,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"201889ff-13d3-4437-b292-6847a6833255"},"execution_count":44,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 52 columns\n","AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRL...SKA COATB_BPIKE/30-81\n","AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKL...SRA Q9T0Q8_BPIKE/1-52\n","DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRL...SKA COATB_BPI22/32-83\n","AEGDDP---AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA COATB_BPM13/24-72\n","AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA COATB_BPZJ2/1-49\n","AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA Q9T0Q9_BPFD/1-49\n","FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL...SRA COATB_BPIF1/22-73\n"]}]},{"cell_type":"code","source":["print(alignment[3:7])                                                           #prints rows from index [3] till [7]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fOfNUJ3szAH1","executionInfo":{"status":"ok","timestamp":1732697133364,"user_tz":-330,"elapsed":492,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"4a4cc4e9-21f3-431b-921e-a3a0880d7302"},"execution_count":24,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 4 rows and 52 columns\n","AEGDDP---AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA COATB_BPM13/24-72\n","AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA COATB_BPZJ2/1-49\n","AEGDDP---AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL...SKA Q9T0Q9_BPFD/1-49\n","FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL...SRA COATB_BPIF1/22-73\n"]}]},{"cell_type":"code","source":["print(alignment[0, 6])                                                          #prints the Amino acid present in the 0th row, 6th column"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Z7mGCEQgzU0T","executionInfo":{"status":"ok","timestamp":1732697230245,"user_tz":-330,"elapsed":492,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"deeaada6-b7c5-46e7-c358-923a73f29a35"},"execution_count":26,"outputs":[{"output_type":"stream","name":"stdout","text":["T\n"]}]},{"cell_type":"code","source":["print(alignment[:, 5])                                                          #prints 5th column"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"K1h__tqpzyBj","executionInfo":{"status":"ok","timestamp":1732697353244,"user_tz":-330,"elapsed":474,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"db783b03-383b-4367-94ee-25237e95169f"},"execution_count":27,"outputs":[{"output_type":"stream","name":"stdout","text":["AAAPPPA\n"]}]},{"cell_type":"code","source":["print(alignment[3:6, :6])                                                       #prints from 3rd row till 6th row but only upto 1st 6 columns"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"dYybEjLM0Iud","executionInfo":{"status":"ok","timestamp":1732697493777,"user_tz":-330,"elapsed":619,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"0292bdb8-38e4-46a1-c093-31e8b7d2f800"},"execution_count":29,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 3 rows and 6 columns\n","AEGDDP COATB_BPM13/24-72\n","AEGDDP COATB_BPZJ2/1-49\n","AEGDDP Q9T0Q9_BPFD/1-49\n"]}]},{"cell_type":"code","source":["print(alignment[:, 6:9])                                                        #prints columns 6 to 9 in all rows"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"C77seAyx0iZb","executionInfo":{"status":"ok","timestamp":1732697579462,"user_tz":-330,"elapsed":482,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"0eadad10-7963-440b-d3e2-d8b8bd84a676"},"execution_count":30,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 3 columns\n","TNY COATB_BPIKE/30-81\n","TNY Q9T0Q8_BPIKE/1-52\n","TSY COATB_BPI22/32-83\n","--- COATB_BPM13/24-72\n","--- COATB_BPZJ2/1-49\n","--- Q9T0Q9_BPFD/1-49\n","TSQ COATB_BPIF1/22-73\n"]}]},{"cell_type":"code","source":["print(alignment[:, 9:])                                                         #prints all rows started from column 9"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jG6qxV810jDj","executionInfo":{"status":"ok","timestamp":1732697582879,"user_tz":-330,"elapsed":492,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"c6924f56-6199-4b9f-825b-0d033fb310a8"},"execution_count":31,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 43 columns\n","ATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA COATB_BPIKE/30-81\n","ATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA Q9T0Q8_BPIKE/1-52\n","ATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA COATB_BPI22/32-83\n","AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA COATB_BPM13/24-72\n","AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA COATB_BPZJ2/1-49\n","AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA Q9T0Q9_BPFD/1-49\n","AKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA COATB_BPIF1/22-73\n"]}]},{"cell_type":"markdown","source":["We can edit the sequenceas per our requirements"],"metadata":{"id":"h_yCesSN05Cz"}},{"cell_type":"code","source":["edited = alignment[:, :6] + alignment[:, 9:]\n","print(edited)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"T95Unu0i03WE","executionInfo":{"status":"ok","timestamp":1732697647107,"user_tz":-330,"elapsed":485,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"7e219132-e1ab-4ad2-fb62-05f6f01a7086"},"execution_count":32,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 49 columns\n","AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA COATB_BPIKE/30-81\n","AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA Q9T0Q8_BPIKE/1-52\n","DGTSTAATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA COATB_BPI22/32-83\n","AEGDDPAKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA COATB_BPM13/24-72\n","AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA COATB_BPZJ2/1-49\n","AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA Q9T0Q9_BPFD/1-49\n","FAADDAAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA COATB_BPIF1/22-73\n"]}]},{"cell_type":"markdown","source":["**Alignment Tools**\n","- create a command line object specifying the options (e.g. the input filename and the output filename)\n","- invoke this command line via a Python operating system call (e.g. using the subprocess module)."],"metadata":{"id":"77W2IQ0W1uQT"}},{"cell_type":"code","source":["import Bio.Align.Applications\n","dir(Bio.Align.Applications)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"trlBkGxK1_C8","executionInfo":{"status":"ok","timestamp":1732697930183,"user_tz":-330,"elapsed":500,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"f1cf86c7-bb1a-4836-8edc-2eb25f8f7475"},"execution_count":35,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/Bio/Application/__init__.py:39: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n","\n","Due to the on going maintenance burden of keeping command line application\n","wrappers up to date, we have decided to deprecate and eventually remove these\n","modules.\n","\n","We instead now recommend building your command line and invoking it directly\n","with the subprocess module.\n","  warnings.warn(\n"]},{"output_type":"execute_result","data":{"text/plain":["['ClustalOmegaCommandline',\n"," 'ClustalwCommandline',\n"," 'DialignCommandline',\n"," 'MSAProbsCommandline',\n"," 'MafftCommandline',\n"," 'MuscleCommandline',\n"," 'PrankCommandline',\n"," 'ProbconsCommandline',\n"," 'TCoffeeCommandline',\n"," '_ClustalOmega',\n"," '_Clustalw',\n"," '_Dialign',\n"," '_MSAProbs',\n"," '_Mafft',\n"," '_Muscle',\n"," '_Prank',\n"," '_Probcons',\n"," '_TCoffee',\n"," '__all__',\n"," '__builtins__',\n"," '__cached__',\n"," '__doc__',\n"," '__file__',\n"," '__loader__',\n"," '__name__',\n"," '__package__',\n"," '__path__',\n"," '__spec__']"]},"metadata":{},"execution_count":35}]},{"cell_type":"markdown","source":["ClustalW"],"metadata":{"id":"t_1vfO4h2HU7"}},{"cell_type":"code","source":["from Bio.Align.Applications import ClustalwCommandline\n","#help(ClustalwCommandline)"],"metadata":{"id":"9TynQ8Qz2H9E","executionInfo":{"status":"ok","timestamp":1732698345402,"user_tz":-330,"elapsed":487,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}}},"execution_count":40,"outputs":[]},{"cell_type":"code","source":["from Bio.Align.Applications import ClustalwCommandline\n","cline = ClustalwCommandline(\"clustalw2\", infile=\"opuntia.fasta\")\n","print(cline)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"FpA1mD2Z2PNc","executionInfo":{"status":"ok","timestamp":1732699260315,"user_tz":-330,"elapsed":504,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"146142ef-bcbf-4463-e289-fb544c311e9f"},"execution_count":56,"outputs":[{"output_type":"stream","name":"stdout","text":["clustalw2 -infile=opuntia.fasta\n"]}]},{"cell_type":"markdown","source":["MUSCLE:\n","- more recent than clustalw"],"metadata":{"id":"0bZXaQhm7Lpe"}},{"cell_type":"code","source":["!apt-get update\n","!apt-get install muscle"],"metadata":{"collapsed":true,"colab":{"base_uri":"https://localhost:8080/"},"id":"6FCi7-ln9YIg","executionInfo":{"status":"ok","timestamp":1732699867601,"user_tz":-330,"elapsed":15424,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"f753d30b-335d-40c7-91ca-f9be76c20290"},"execution_count":68,"outputs":[{"output_type":"stream","name":"stdout","text":["Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]\n","Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n","Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease\n","Get:4 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n","Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n","Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n","Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n","Get:8 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,506 kB]\n","Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n","Get:10 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n","Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n","Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [2,738 kB]\n","Get:13 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,224 kB]\n","Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,513 kB]\n","Get:15 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,454 kB]\n","Get:16 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [33.8 kB]\n","Get:17 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,619 kB]\n","Fetched 19.5 MB in 4s (4,924 kB/s)\n","Reading package lists... Done\n","W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n","Reading package lists... Done\n","Building dependency tree... Done\n","Reading state information... Done\n","The following NEW packages will be installed:\n","  muscle\n","0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n","Need to get 244 kB of archives.\n","After this operation, 709 kB of additional disk space will be used.\n","Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 muscle amd64 1:3.8.1551-2build1 [244 kB]\n","Fetched 244 kB in 2s (153 kB/s)\n","Selecting previously unselected package muscle.\n","(Reading database ... 123630 files and directories currently installed.)\n","Preparing to unpack .../muscle_1%3a3.8.1551-2build1_amd64.deb ...\n","Unpacking muscle (1:3.8.1551-2build1) ...\n","Setting up muscle (1:3.8.1551-2build1) ...\n","Processing triggers for man-db (2.10.2-1) ...\n"]}]},{"cell_type":"code","source":["from Bio.Align.Applications import MuscleCommandline\n","muscle_cline = MuscleCommandline(input=\"opuntia.fasta\")\n","print(muscle_cline)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5s-OkHW67T-O","executionInfo":{"status":"ok","timestamp":1732699874256,"user_tz":-330,"elapsed":487,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"83ecaf92-238a-4951-de9a-a4ebc08eb4fa"},"execution_count":69,"outputs":[{"output_type":"stream","name":"stdout","text":["muscle -in opuntia.fasta\n"]}]},{"cell_type":"code","source":["from io import StringIO\n","muscle_cline = MuscleCommandline(input=\"opuntia.fasta\")\n","stdout, stderr = muscle_cline()                                                 #output is a string, StringIO is used to turn it into a handle\n","\n","from Bio import AlignIO\n","align = AlignIO.read(StringIO(stdout), \"fasta\")\n","print(align)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"IQTJEXDF8SL2","executionInfo":{"status":"ok","timestamp":1732699981376,"user_tz":-330,"elapsed":745,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"f274a04d-88f3-4db4-d0ba-e20f87b237d4"},"execution_count":71,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 906 columns\n","TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191663\n","TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191665\n","TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191664\n","TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191661\n","TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191660\n","TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191659\n","TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191658\n"]}]},{"cell_type":"code","source":["#Using the subprocess module we can work directly with handles instead\n","import subprocess\n","import sys                                                                      # Import the sys module\n","from Bio.Align.Applications import MuscleCommandline\n","muscle_cline = MuscleCommandline(input=\"opuntia.fasta\")\n","child = subprocess.Popen(str(muscle_cline),\n"," stdout=subprocess.PIPE,\n"," stderr=subprocess.PIPE,\n"," universal_newlines=True,\n"," shell=(sys.platform!=\"win32\"))\n","\n","from Bio import AlignIO\n","align = AlignIO.read(child.stdout, \"fasta\")\n","print(align)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Y86mWDDS-i-e","executionInfo":{"status":"ok","timestamp":1732700248110,"user_tz":-330,"elapsed":983,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"af9c8732-0376-4561-8c51-fbad70c585c8"},"execution_count":73,"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment with 7 rows and 906 columns\n","TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191663\n","TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191665\n","TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191664\n","TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191661\n","TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191660\n","TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191659\n","TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191658\n"]}]},{"cell_type":"markdown","source":["**EMBOSS needle and water**"],"metadata":{"id":"7d0ZG6My_ESu"}},{"cell_type":"markdown","source":["Global pairwise alignment between two sequences, prepared in FASTA format as follows:"],"metadata":{"id":"Gt90XHRr_2WP"}},{"cell_type":"code","source":["!apt-get update\n","!apt-get install emboss"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PsAnzJWWCDZY","executionInfo":{"status":"ok","timestamp":1732701115787,"user_tz":-330,"elapsed":37325,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"76cc9557-ce0c-4079-aab5-cc447758733b"},"execution_count":84,"outputs":[{"output_type":"stream","name":"stdout","text":["\r0% [Working]\r            \rHit:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n","Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n","Hit:3 http://security.ubuntu.com/ubuntu jammy-security InRelease\n","Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease\n","Hit:5 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n","Hit:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n","Hit:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n","Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n","Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n","Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n","Reading package lists... Done\n","W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n","Reading package lists... Done\n","Building dependency tree... Done\n","Reading state information... Done\n","The following additional packages will be installed:\n","  dialign emboss-data emboss-lib libhpdf-2.3.0 primer3\n","Suggested packages:\n","  emboss-doc emboss-test embassy clustalw ncbi-epcr\n","The following NEW packages will be installed:\n","  dialign emboss emboss-data emboss-lib libhpdf-2.3.0 primer3\n","0 upgraded, 6 newly installed, 0 to remove and 49 not upgraded.\n","Need to get 65.7 MB of archives.\n","After this operation, 500 MB of additional disk space will be used.\n","Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 dialign amd64 2.2.1-11 [139 kB]\n","Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhpdf-2.3.0 amd64 2.3.0+dfsg-1build1 [335 kB]\n","Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 emboss-lib amd64 6.6.0+dfsg-11ubuntu1 [2,971 kB]\n","Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 emboss-data all 6.6.0+dfsg-11ubuntu1 [61.0 MB]\n","Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 emboss amd64 6.6.0+dfsg-11ubuntu1 [977 kB]\n","Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 primer3 amd64 2.5.0-1 [208 kB]\n","Fetched 65.7 MB in 7s (9,430 kB/s)\n","Selecting previously unselected package dialign.\n","(Reading database ... 123639 files and directories currently installed.)\n","Preparing to unpack .../0-dialign_2.2.1-11_amd64.deb ...\n","Unpacking dialign (2.2.1-11) ...\n","Selecting previously unselected package libhpdf-2.3.0:amd64.\n","Preparing to unpack .../1-libhpdf-2.3.0_2.3.0+dfsg-1build1_amd64.deb ...\n","Unpacking libhpdf-2.3.0:amd64 (2.3.0+dfsg-1build1) ...\n","Selecting previously unselected package emboss-lib.\n","Preparing to unpack .../2-emboss-lib_6.6.0+dfsg-11ubuntu1_amd64.deb ...\n","Unpacking emboss-lib (6.6.0+dfsg-11ubuntu1) ...\n","Selecting previously unselected package emboss-data.\n","Preparing to unpack .../3-emboss-data_6.6.0+dfsg-11ubuntu1_all.deb ...\n","Unpacking emboss-data (6.6.0+dfsg-11ubuntu1) ...\n","Selecting previously unselected package emboss.\n","Preparing to unpack .../4-emboss_6.6.0+dfsg-11ubuntu1_amd64.deb ...\n","Unpacking emboss (6.6.0+dfsg-11ubuntu1) ...\n","Selecting previously unselected package primer3.\n","Preparing to unpack .../5-primer3_2.5.0-1_amd64.deb ...\n","Unpacking primer3 (2.5.0-1) ...\n","Setting up libhpdf-2.3.0:amd64 (2.3.0+dfsg-1build1) ...\n","Setting up emboss-data (6.6.0+dfsg-11ubuntu1) ...\n","Setting up primer3 (2.5.0-1) ...\n","Setting up emboss-lib (6.6.0+dfsg-11ubuntu1) ...\n","Setting up dialign (2.2.1-11) ...\n","Setting up emboss (6.6.0+dfsg-11ubuntu1) ...\n","Processing triggers for man-db (2.10.2-1) ...\n","Processing triggers for shared-mime-info (2.1-2) ...\n","Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n","/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n","\n"]}]},{"cell_type":"code","source":["#Perform needlemann wunsch alignment\n","from Bio.Emboss.Applications import NeedleCommandline                           #commandline for needlemann wunsch algo\n","needle_cline = NeedleCommandline()\n","needle_cline.asequence=\"alpha.faa\"\n","needle_cline.bsequence=\"beta.faa\"\n","needle_cline.gapopen=10\n","needle_cline.gapextend=0.5\n","needle_cline.outfile=\"needle.txt\"\n","print(needle_cline)\n","print(needle_cline.outfile)\n","\n","stdout, stderr = needle_cline()\n","print(stdout + stderr)\n","\n","from Bio import AlignIO\n","align = AlignIO.read(\"needle.txt\", \"emboss\")                                    #writes the output file\n","print(align)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Y68kNQs2CX03","executionInfo":{"status":"ok","timestamp":1732701409766,"user_tz":-330,"elapsed":656,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"4409f753-4fa9-4c2c-dc21-3bddfd2a29af"},"execution_count":91,"outputs":[{"output_type":"stream","name":"stdout","text":["needle -outfile=needle.txt -asequence=alpha.faa -bsequence=beta.faa -gapopen=10 -gapextend=0.5\n","needle.txt\n","Needleman-Wunsch global alignment of two sequences\n","\n","Alignment with 2 rows and 149 columns\n","MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTY...KYR HBA_HUMAN\n","MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRF...KYH HBB_HUMAN\n"]}]},{"cell_type":"code","source":["# Perform Smith-Waterman alignment (local)\n","water_cline = WaterCommandline(asequence=asequence, bsequence=bsequence,\n","                                gapopen=10, gapextend=0.5, outfile=\"water.txt\")\n","stdout, stderr = water_cline()\n","\n","# Check for errors in stderr\n","if stderr:\n","    print(\"Water Error:\", stderr)\n","else:\n","    print(\"Water alignment completed successfully.\")\n","\n","# Read the local alignment from the output file\n","try:\n","    water_alignment = AlignIO.read(\"water.txt\", \"emboss\")\n","    print(\"Water Alignment:\")\n","    print(water_alignment)\n","except FileNotFoundError as e:\n","    print(f\"File not found: {e}\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"968Ar2c9CkJG","executionInfo":{"status":"ok","timestamp":1732701367674,"user_tz":-330,"elapsed":487,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"8f722144-ad8b-4ab8-baa5-5c1191681966"},"execution_count":90,"outputs":[{"output_type":"stream","name":"stdout","text":["Water Error: Smith-Waterman local alignment of sequences\n","\n","Water Alignment:\n","Alignment with 2 rows and 145 columns\n","LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPH...SKY HBA_HUMAN\n","LTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFES...HKY HBB_HUMAN\n"]}]},{"cell_type":"markdown","source":["**PAIRWISE ALIGNMENT**"],"metadata":{"id":"YrKzIWKtFzSI"}},{"cell_type":"code","source":["from Bio import pairwise2\n","from Bio import SeqIO\n","\n","# Read sequences from FASTA files\n","seq1 = SeqIO.read(\"alpha.faa\", \"fasta\").seq\n","seq2 = SeqIO.read(\"beta.faa\", \"fasta\").seq\n","\n","# Perform global alignment\n","alignments = pairwise2.align.globalxx(seq1, seq2)\n","\n","# Print the first alignment in a formatted way\n","print(pairwise2.format_alignment(*alignments[0]))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SefTPZKoGFT4","executionInfo":{"status":"ok","timestamp":1732702135448,"user_tz":-330,"elapsed":489,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"dedae4ec-8796-4ba6-b9a3-0bc2a3e84b30"},"execution_count":106,"outputs":[{"output_type":"stream","name":"stdout","text":["MV-LSPADKTNV---K-A--A-WGKVGAHAG---EY-GA-EALE-RMFLSF----PTTK-TY--FPHFDL-SH-G--S---AQVK-G------HGKKV--A--DA-LTNAVAHVD-DMPNALS----A-LSD-LHAH--KLR-VDPV-NFKL-LSHCL---LVT--LAAHLPA----EFTPA-VH-ASLDKFLAS---VSTV------LTS--KYR-\n","|| |     |     | |  | ||||        |  |  |||  |  |      |    |   |  |   |  |  |   | |  |      |||||  |  |  |    ||   |  | |     | ||  |  |  ||  |||  ||   |   |   ||   | ||       ||||  |  |      |    |  |      |    ||  \n","MVHL-----T--PEEKSAVTALWGKV-----NVDE-VG-GEAL-GR--L--LVVYP---WT-QRF--F--ES-FGDLSTPDA-V-MGNPKVKAHGKKVLGAFSD-GL----AH--LD--N-L-KGTFATLS-EL--HCDKL-HVDP-ENF--RL---LGNVLV-CVL-AH---HFGKEFTP-PV-QA------A-YQKV--VAGVANAL--AHKY-H\n","  Score=72\n","\n"]}]},{"cell_type":"markdown","source":["**BLOSUM62 matrix**"],"metadata":{"id":"W2qIlbeHE90g"}},{"cell_type":"code","source":["!pip install biopython --upgrade"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"69mj99PlFNHf","executionInfo":{"status":"ok","timestamp":1732701910087,"user_tz":-330,"elapsed":5019,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"0cb0e791-3f99-448c-b912-f20b0c8398f0"},"execution_count":101,"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: biopython in /usr/local/lib/python3.10/dist-packages (1.84)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.26.4)\n"]}]},{"cell_type":"code","source":["from Bio import pairwise2\n","from Bio import SeqIO\n","from Bio.Align import substitution_matrices                                     # Import substitution_matrices\n","seq1 = SeqIO.read(\"alpha.faa\", \"fasta\")\n","seq2 = SeqIO.read(\"beta.faa\", \"fasta\")\n","blosum62 = substitution_matrices.load(\"BLOSUM62\")                               # Load blosum62 matrix\n","alignments = pairwise2.align.globalds(seq1.seq, seq2.seq, blosum62, -10, -0.5)\n","len(alignments)\n","\n","\n","print(pairwise2.format_alignment(*alignments[0]))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"EVZp9PSVFAY_","executionInfo":{"status":"ok","timestamp":1732702028676,"user_tz":-330,"elapsed":494,"user":{"displayName":"PRIYADHARSHINI� .T","userId":"17953581082584727495"}},"outputId":"3cf9455a-d911-4a64-fbfd-39711decda37"},"execution_count":105,"outputs":[{"output_type":"stream","name":"stdout","text":["MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS-----HGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR\n","|| |.|..|..|.|.||||  ...|.|.|||.|.....|.|...|..| |||     .|...||.|||||..|.....||.|........||.||..||.|||.||.||...|...||.|...||||.|.|...|..|.|...|..||.\n","MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH\n","  Score=292.5\n","\n"]}]}]}